examples/mnist/mnist-common.h

```diff
@@ -44,28 +52,41 @@ struct mnist_model {
     static const size_t size_weight  = 100 *      1024*1024;
     static const size_t size_compute =   1 * 1024*1024*1024;
 
-    void                * buf_weight  = nullptr;
     struct ggml_context * ctx_weight  = nullptr;
-    void                * buf_compute = nullptr;
     struct ggml_context * ctx_compute = nullptr;
+    ggml_backend_buffer_t buf_weight  = nullptr;
+    ggml_backend_buffer_t buf_compute = nullptr;
+
+    mnist_model(const std::string & backend_name) {
+        const size_t backend_index = ggml_backend_reg_find_by_name(backend_name.c_str());
+        if (backend_index == SIZE_MAX) {
+            fprintf(stderr, "%s: ERROR: backend %s not found, available:\n", __func__, backend_name.c_str());
+            for (size_t i = 0; i < ggml_backend_reg_get_count(); ++i) {
+                fprintf(stderr, "  - %s\n", ggml_backend_reg_get_name(i));
+            }
+            exit(1);
+        }
+
+        fprintf(stderr, "%s: using %s backend\n", __func__, backend_name.c_str());
+        backend = ggml_backend_reg_init_backend(backend_index, nullptr);
+        if (ggml_backend_is_cpu(backend)) {
+            ggml_backend_cpu_set_n_threads(backend, std::thread::hardware_concurrency());
+        }
 
-    mnist_model() {
-        buf_weight = malloc(size_weight);
         {
             struct ggml_init_params params = {
                 /*.mem_size   =*/ size_weight,
-                /*.mem_buffer =*/ buf_weight,
-                /*.no_alloc   =*/ false,
+                /*.mem_buffer =*/ nullptr,
+                /*.no_alloc   =*/ true,
             };
             ctx_weight = ggml_init(params);
         }
 
-        buf_compute = malloc(size_compute);
         {
             struct ggml_init_params params = {
                 /*.mem_size   =*/ size_compute,
-                /*.mem_buffer =*/ buf_compute,
-                /*.no_alloc   =*/ false,
+                /*.mem_buffer =*/ nullptr,
+                /*.no_alloc   =*/ true,
             };
             ctx_compute = ggml_init(params);
         }
```

Comment:
Shouldn't `size_compute` and `size_weight` should be reduced now? this is allocating very big buffers for what looks like just for the metadata.

Reply:
Thanks, I forgot.